libs/community/langchain_community/graph_vectorstores/extractors/link_extractor_transformer.py

```diff
@@ -0,0 +1,44 @@
+from typing import Any, Iterable, Sequence
+
+from langchain_core.documents import Document
+from langchain_core.documents.transformers import BaseDocumentTransformer
+from langchain_core.graph_vectorstores.links import add_links
+
+from langchain_community.graph_vectorstores.extractors.link_extractor import (
+    LinkExtractor,
+)
+
+
+class LinkExtractorTransformer(BaseDocumentTransformer):
+    """DocumentTransformer for applying one or more LinkExtractors.
+
+    Example:
+        .. code-block:: python
+
+            extract_links = LinkExtractorTransformer([
+                HtmlLinkExtractor().as_document_extractor(),
+            ])
+            extract_links.transform_documents(docs)
+    """
+
+    def __init__(self, link_extractors: Iterable[LinkExtractor[Document]]):
+        """Create a DocumentTransformer which adds extracted links to each document."""
+        self.link_extractors = link_extractors
+
+    def transform_documents(
+        self, documents: Sequence[Document], **kwargs: Any
+    ) -> Sequence[Document]:
+        # Implement `transform_docments` directly, so that LinkExtractors which operate
+        # better in batch (`extract_many`) get a chance to do so.
+        document_links = zip(
+            documents,
+            zip(
+                *[
+                    extractor.extract_many(documents)
+                    for extractor in self.link_extractors
+                ]
+            ),
+        )
+        for document, links in document_links:
+            add_links(document, *links)
```

Comment:
LinkExtractorTransformer violates standard conventions in pythons of having no return type if the input is being mutated.

Users will generally not expect the input documents to be mutated as a result of LinkExtractorTransformer.

You could make this more explicit by adding an opt-in  parameter in the init to do in place mutation

Reply:
I think it's probably better to skip the parameter and just duplicate things. Although, it's somewhat inconsistent within the codebase:

1. I based this on `doctran_text_extract` which mutates the documents in place:

https://github.com/langchain-ai/langchain/blob/cacdf96f9c2a608dfb6c69ca07a585d7cad7786d/libs/community/langchain_community/document_transformers/doctran_text_extract.py#L113

Also, beautiful_soup_transformer:

https://github.com/langchain-ai/langchain/blob/cacdf96f9c2a608dfb6c69ca07a585d7cad7786d/libs/community/langchain_community/document_transformers/beautiful_soup_transformer.py#L76

2. I see many things do what `google_translate` does, which would create a new document (and new metadata):

https://github.com/langchain-ai/langchain/blob/cacdf96f9c2a608dfb6c69ca07a585d7cad7786d/libs/community/langchain_community/document_transformers/google_translate.py#L104

3. Many of the others are somewhere in the middle -- making a new document but passing in the original metadata (so future mutations would affect both documents) or creating a shallow copy of metadata (any collections within the metadata would be mutated).

I think I'll change this to follow the google translate pattern (shallow copy of metadata) since that seems more expected (in general), although it seems like the document transformers are in a generally confusing state.